packages/schematics/angular/migrations/update-ssr-imports/migration.ts

```diff
@@ -0,0 +1,103 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import { DirEntry, Rule, UpdateRecorder } from '@angular-devkit/schematics';
+import * as ts from '../../third_party/github.com/Microsoft/TypeScript/lib/typescript';
+
+function* visit(directory: DirEntry): IterableIterator<ts.SourceFile> {
+  for (const path of directory.subfiles) {
+    if (path.endsWith('.ts') && !path.endsWith('.d.ts')) {
+      const entry = directory.file(path);
+      if (entry) {
+        const content = entry.content;
+        if (content.includes('CommonEngine') && !content.includes('@angular/ssr/node')) {
+          const source = ts.createSourceFile(
+            entry.path,
+            content.toString().replace(/^\uFEFF/, ''),
+            ts.ScriptTarget.Latest,
+            true,
+          );
+
+          yield source;
+        }
+      }
+    }
+  }
+
+  for (const path of directory.subdirs) {
+    if (path === 'node_modules' || path.startsWith('.')) {
+      continue;
+    }
+
+    yield* visit(directory.dir(path));
+  }
+}
+
+/**
+ * Schematics rule that identifies and updates import declarations in TypeScript files.
+ * Specifically, it modifies imports of '@angular/ssr' by appending '/node' if the
+ * `CommonEngine` is used from the old entry point.
+ *
+ */
+export default function (): Rule {
+  return (tree) => {
+    for (const sourceFile of visit(tree.root)) {
+      let recorder: UpdateRecorder | undefined;
+
+      const allImportDeclarations = sourceFile.statements.filter((n) => ts.isImportDeclaration(n));
+      if (allImportDeclarations.length === 0) {
+        continue;
+      }
+
+      const ssrImports = filterImportDeclaration('@angular/ssr', allImportDeclarations);
+      if (!ssrImports) {
+        continue;
+      }
+
+      for (const ssrImport of ssrImports) {
+        const ssrNamedBinding = getNamedImports(ssrImport);
+        if (ssrNamedBinding) {
+          const isUsingOldEntryPoint = ssrNamedBinding.elements.some((e) =>
+            e.name.text.startsWith('CommonEngine'),
+          );
+
+          if (!isUsingOldEntryPoint) {
+            continue;
+          }
+
+          recorder ??= tree.beginUpdate(sourceFile.fileName);
+          recorder.insertRight(ssrImport.moduleSpecifier.getEnd() - 1, '/node');
```

Comment:
There are no other exported symbols

Reply:
Ah cool, that makes this easier then. If so, do we really care about `CommonEngine` at all? Should we just look for imports of `@angular/ssr` and replace them with `@angular/ssr/node`?

I think an unused `import {} from '@angular/ssr';` or `import * as ssr from '@angular/ssr';` might fall through the cracks here.

Reply:
We need to check that we only migrate when the symbols start with `CommonEngine` otherwise the migration will not be idempotent.

Reply:
IMHO, those 2 cases are rare enough that doesn’t warrant to be included.

Reply:
> We need to check that we only migrate when the symbols start with `CommonEngine` otherwise the migration will not be idempotent.

I don't see why an `@angular/ssr` -> `@angular/ssr/node` rename wouldn't be idempotent, but I'll defer to you that's it's necessary.

I agree the `import {}` and `import * as ssr` cases probably aren't important enough to go out of our way for.